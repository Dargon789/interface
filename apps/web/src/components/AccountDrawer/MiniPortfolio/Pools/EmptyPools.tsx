import { useCallback } from 'react'
import { useTranslation } from 'react-i18next'
import { useNavigate } from 'react-router'
import { Button } from 'ui/src'
import { EmptyPoolsIcon } from 'ui/src/components/icons/EmptyPoolsIcon'
import { Flex } from 'ui/src/components/layout'
import { Text } from 'ui/src/components/text'

export function EmptyPools({
  onNavigateClick,
  hasSolanaAndEVMWalletsConnected,
}: {
  onNavigateClick?: () => void
  hasSolanaAndEVMWalletsConnected?: boolean
}) {
  const { t } = useTranslation()
  const navigate = useNavigate()

  const actionButtonClick = useCallback(() => {
    onNavigateClick?.()
    navigate('/positions')
  }, [navigate, onNavigateClick])

  return (
    <Flex
      alignItems="center"
      justifyContent="center"
      height="100%"
      width="100%"
      paddingLeft="$spacing32"
      paddingRight="$spacing32"
      $sm={{ paddingTop: '$spacing8' }}
    >
      <EmptyPoolsIcon size={115} />
      <Text variant="subheading1" textAlign="center" marginTop="$spacing12">
        {t('nft.noPools')}
      </Text>
      <Text
        variant="body2"
        textAlign="center"
        marginTop="$spacing8"
        color="$neutral2"
        $platform-web={{ textWrap: 'pretty' }}
      >
        {hasSolanaAndEVMWalletsConnected ? t('pool.openToStart.evmAndSolanaConnected') : t('pool.openToStart.evmOnly')}
      </Text>
      <Flex marginTop="$spacing20" row>
        <Button variant="branded" onPress={actionButtonClick} size="small">
          {t('pool.newPosition.plus')}
        </Button>
      </Flex>
    </Flex>
  )
}
