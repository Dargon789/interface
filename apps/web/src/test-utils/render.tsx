import { AssetActivityProvider } from 'appGraphql/data/apollo/AssetActivityProvider'
import { TokenBalancesProvider } from 'appGraphql/data/apollo/TokenBalancesProvider'
import { MockedProvider } from '@apollo/client/testing'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { queries } from '@testing-library/dom'
import { RenderOptions, render } from '@testing-library/react'
import { RenderHookOptions, renderHook, WrapperComponent } from '@testing-library/react-hooks'
import { Web3ProviderUpdater } from 'components/Web3Provider'
import TestWeb3Provider from 'components/Web3Provider/TestWeb3Provider'
import { WebUniswapProvider } from 'components/Web3Provider/WebUniswapContext'
import { WebAccountsStoreProvider } from 'features/accounts/store/provider'
import { ConnectWalletMutationProvider } from 'features/wallet/connection/hooks/useConnectWalletMutation'
import { ExternalWalletProvider } from 'features/wallet/providers/ExternalWalletProvider'
import { BlockNumberContext } from 'lib/hooks/useBlockNumber'
import { PropsWithChildren, ReactElement, ReactNode } from 'react'
import { HelmetProvider } from 'react-helmet-async/lib/index'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router'
import store from 'state'
import { ThemeProvider } from 'theme'
import { TamaguiProvider } from 'theme/tamaguiProvider'
import { ReactRouterUrlProvider } from 'uniswap/src/contexts/UrlContext'
import { MismatchContextProvider } from 'uniswap/src/features/smartWallet/mismatch/MismatchContext'

const queryClient = new QueryClient()

const BLOCK_NUMBER_CONTEXT = { fastForward: () => {}, block: 1234, mainnetBlock: 1234 }
function MockedBlockNumberProvider({ children }: PropsWithChildren) {
  return <BlockNumberContext.Provider value={BLOCK_NUMBER_CONTEXT}>{children}</BlockNumberContext.Provider>
}

function MockedMismatchProvider({ children }: PropsWithChildren) {
  return (
    <MismatchContextProvider
      address={undefined}
      chainId={undefined}
      mismatchCallback={() => Promise.resolve({ [String(1)]: false })}
      onHasAnyMismatch={() => {}}
      chains={[1]}
      defaultChainId={1}
      isTestnetModeEnabled={false}
    >
      {children}
    </MismatchContextProvider>
  )
}

function CommonTestProviders({ children }: PropsWithChildren) {
  return (
    <MockedProvider showWarnings={false}>
      <AssetActivityProvider>
        <TokenBalancesProvider>
          <ReactRouterUrlProvider>
            <MockedBlockNumberProvider>
              <ThemeProvider>
                <TamaguiProvider>
                  <Web3ProviderUpdater />
                  <MockedMismatchProvider>{children}</MockedMismatchProvider>
                </TamaguiProvider>
              </ThemeProvider>
            </MockedBlockNumberProvider>
          </ReactRouterUrlProvider>
        </TokenBalancesProvider>
      </AssetActivityProvider>
    </MockedProvider>
  )
}

function BaseWrapper({
  children,
  includeUniswapContext = false,
}: PropsWithChildren<{ includeUniswapContext?: boolean }>) {
  return (
    <HelmetProvider>
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          <BrowserRouter>
            <TestWeb3Provider>
              <ConnectWalletMutationProvider>
                <WebAccountsStoreProvider>
                  <ExternalWalletProvider>
                    {/* TODO: figure out how to properly mock `WebUniswapProvider` so that we can include it in all tests */}
                    {includeUniswapContext ? (
                      <WebUniswapProvider>
                        <CommonTestProviders>{children}</CommonTestProviders>
                      </WebUniswapProvider>
                    ) : (
                      <CommonTestProviders>{children}</CommonTestProviders>
                    )}
                  </ExternalWalletProvider>
                </WebAccountsStoreProvider>
              </ConnectWalletMutationProvider>
            </TestWeb3Provider>
          </BrowserRouter>
        </QueryClientProvider>
      </Provider>
    </HelmetProvider>
  )
}

const WithProviders = ({ children }: { children?: ReactNode }) => <BaseWrapper>{children}</BaseWrapper>
const WithUniswapProviders = ({ children }: { children?: ReactNode }) => (
  <BaseWrapper includeUniswapContext>{children}</BaseWrapper>
)

type CustomRenderOptions = Omit<RenderOptions, 'wrapper'>

const customRender = (ui: ReactElement, options?: CustomRenderOptions) => {
  return render<typeof queries>(ui, { ...options, wrapper: WithProviders })
}

const customRenderWithUniswapContext = (ui: ReactElement, options?: CustomRenderOptions) => {
  return render<typeof queries>(ui, { ...options, wrapper: WithUniswapProviders })
}

type CustomRenderHookOptions<Props> = Omit<RenderHookOptions<Props>, 'wrapper'>
const customRenderHook = <Result, Props>(
  hook: (initialProps: Props) => Result,
  options?: CustomRenderHookOptions<Props>,
) => {
  return renderHook(hook, { ...options, wrapper: WithProviders as WrapperComponent<Props> })
}

// Testing utils may export *.
// eslint-disable-next-line no-restricted-syntax
export * from '@testing-library/react'
export {
  customRender as render,
  customRenderWithUniswapContext as renderWithUniswapContext,
  customRenderHook as renderHook,
}
