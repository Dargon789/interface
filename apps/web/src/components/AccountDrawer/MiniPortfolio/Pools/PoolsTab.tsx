import { PositionStatus, ProtocolVersion } from '@uniswap/client-pools/dist/pools/v1/types_pb'
import { ExpandoRow } from 'components/AccountDrawer/MiniPortfolio/ExpandoRow'
import { useAccountDrawer } from 'components/AccountDrawer/MiniPortfolio/hooks'
import { EmptyPools } from 'components/AccountDrawer/MiniPortfolio/Pools/EmptyPools'
import { PortfolioSkeleton } from 'components/AccountDrawer/MiniPortfolio/PortfolioRow'
import { SolanaOnlyEmptyState } from 'components/AccountDrawer/MiniPortfolio/SolanaOnlyEmptyState'
import { LiquidityPositionCard } from 'components/Liquidity/LiquidityPositionCard'
import { PositionInfo } from 'components/Liquidity/types'
import { getPositionUrl } from 'components/Liquidity/utils/getPositionUrl'
import { parseRestPosition } from 'components/Liquidity/utils/parseFromRest'
import { useAccount } from 'hooks/useAccount'
import { useSwitchChain } from 'hooks/useSwitchChain'
import { useMemo, useReducer, useState } from 'react'
import { useTranslation } from 'react-i18next'
import { useNavigate } from 'react-router'
import { AnimatePresence, Flex, TouchableArea } from 'ui/src'
import { zIndexes } from 'ui/src/theme'
import { ZERO_ADDRESS } from 'uniswap/src/constants/misc'
import { useGetPositionsQuery } from 'uniswap/src/data/rest/getPositions'
import { useActiveAddresses } from 'uniswap/src/features/accounts/store/hooks'
import { useEnabledChains } from 'uniswap/src/features/chains/hooks/useEnabledChains'
import { ElementName } from 'uniswap/src/features/telemetry/constants'
import Trace from 'uniswap/src/features/telemetry/Trace'
import { usePositionVisibilityCheck } from 'uniswap/src/features/visibility/hooks/usePositionVisibilityCheck'
import { useEvent } from 'utilities/src/react/hooks'

function isPositionInfo(position: PositionInfo | undefined): position is PositionInfo {
  return !!position
}

function getPositionKey(position: PositionInfo) {
  const { chainId } = position
  if (position.version === ProtocolVersion.V2) {
    return `${position.liquidityToken.address}-${chainId}`
  }

  return `${position.tokenId}-${chainId}`
}

export default function Pools({ account }: { account: string }) {
  const { t } = useTranslation()
  const { chains } = useEnabledChains()
  const isPositionVisible = usePositionVisibilityCheck()
  const { evmAddress, svmAddress } = useActiveAddresses()
  const accountDrawer = useAccountDrawer()
  const [showClosed, toggleShowClosed] = useReducer((showClosed) => !showClosed, false)
  const [showHidden, setShowHidden] = useState(false)
  const [openMenuPositionKey, setOpenMenuPositionKey] = useState<string | null>(null)

  // Skip queries if Solana-only wallet
  const skipQueries = Boolean(svmAddress && !evmAddress)

  const { data, isLoading } = useGetPositionsQuery(
    {
      address: account,
      chainIds: chains,
      positionStatuses: [PositionStatus.IN_RANGE, PositionStatus.OUT_OF_RANGE],
      protocolVersions: [ProtocolVersion.V2, ProtocolVersion.V3, ProtocolVersion.V4],
      includeHidden: true,
    },
    skipQueries,
  )

  const { data: closedData } = useGetPositionsQuery(
    {
      address: account,
      chainIds: chains,
      positionStatuses: [PositionStatus.CLOSED],
      protocolVersions: [ProtocolVersion.V2, ProtocolVersion.V3, ProtocolVersion.V4],
      includeHidden: true,
    },
    skipQueries,
  )

  const openPositions = useMemo(() => {
    if (skipQueries) {
      return undefined
    }
    return data?.positions.map(parseRestPosition).filter(isPositionInfo)
  }, [data?.positions, skipQueries])

  const closedPositions = useMemo(() => {
    if (skipQueries) {
      return undefined
    }
    return closedData?.positions.map(parseRestPosition).filter(isPositionInfo)
  }, [closedData?.positions, skipQueries])

  const { visibleOpenPositions, visibleClosedPositions, hiddenPositions } = useMemo(() => {
    function splitByVisibility(positions: PositionInfo[] = []) {
      const visible: PositionInfo[] = []
      const hidden: PositionInfo[] = []
      positions.forEach((pos) => {
        if (
          isPositionVisible({
            poolId: pos.poolId,
            tokenId: pos.tokenId,
            chainId: pos.chainId,
            isFlaggedSpam: pos.isHidden,
          })
        ) {
          visible.push(pos)
        } else {
          hidden.push(pos)
        }
      })
      return { visible, hidden }
    }

    const { visible: visibleOpenPositions, hidden: hiddenOpenPositions } = splitByVisibility(openPositions)
    const { visible: visibleClosedPositions, hidden: hiddenClosedPositions } = splitByVisibility(closedPositions)

    return {
      visibleOpenPositions,
      visibleClosedPositions,
      hiddenPositions: [...hiddenOpenPositions, ...hiddenClosedPositions],
    }
  }, [openPositions, closedPositions, isPositionVisible])

  // If Solana-only wallet, show Solana-only empty state with EVM connect button
  if (skipQueries) {
    return <SolanaOnlyEmptyState tab="pools" showConnectButton />
  }

  if (!openPositions && isLoading) {
    return <PortfolioSkeleton />
  }

  if (!openPositions || (openPositions.length === 0 && closedPositions?.length === 0)) {
    // EVM and Solana wallets connected but no pools - Show EmptyPools with dual wallet message
    if (svmAddress && evmAddress) {
      return <EmptyPools onNavigateClick={accountDrawer.close} hasSolanaAndEVMWalletsConnected />
    }

    // EVM-only with no pools
    return <EmptyPools onNavigateClick={accountDrawer.close} />
  }

  return (
    <AnimatePresence>
      <Flex gap="$spacing12">
        {visibleOpenPositions.map((positionInfo) => (
          <PositionListItem
            key={getPositionKey(positionInfo)}
            positionInfo={positionInfo}
            openMenuPositionKey={openMenuPositionKey}
            setOpenMenuPositionKey={setOpenMenuPositionKey}
          />
        ))}
      </Flex>
      {visibleClosedPositions.length > 0 && (
        <ExpandoRow
          title={t('liquidityPool.positions.closed.title')}
          isExpanded={showClosed}
          toggle={toggleShowClosed}
          numItems={visibleClosedPositions.length}
        >
          <Flex gap="$spacing12">
            {visibleClosedPositions.map((positionInfo) => (
              <PositionListItem
                key={getPositionKey(positionInfo)}
                positionInfo={positionInfo}
                openMenuPositionKey={openMenuPositionKey}
                setOpenMenuPositionKey={setOpenMenuPositionKey}
              />
            ))}
          </Flex>
        </ExpandoRow>
      )}
      {hiddenPositions.length > 0 && (
        <ExpandoRow
          title={t('common.hidden')}
          isExpanded={showHidden}
          toggle={() => setShowHidden((prev) => !prev)}
          numItems={hiddenPositions.length}
        >
          <Flex gap="$spacing12">
            {hiddenPositions.map((position) => (
              <PositionListItem
                key={getPositionKey(position)}
                positionInfo={position}
                isVisible={false}
                openMenuPositionKey={openMenuPositionKey}
                setOpenMenuPositionKey={setOpenMenuPositionKey}
              />
            ))}
          </Flex>
        </ExpandoRow>
      )}
    </AnimatePresence>
  )
}

function PositionListItem({
  positionInfo,
  isVisible = true,
  openMenuPositionKey,
  setOpenMenuPositionKey,
}: {
  positionInfo: PositionInfo
  isVisible?: boolean
  openMenuPositionKey: string | null
  setOpenMenuPositionKey: (key: string | null) => void
}) {
  const { chainId, currency0Amount, currency1Amount } = positionInfo
  const token0 = currency0Amount.currency
  const token1 = currency1Amount.currency

  const navigate = useNavigate()
  const accountDrawer = useAccountDrawer()
  const account = useAccount()
  const switchChain = useSwitchChain()
  const positionUrl = getPositionUrl(positionInfo)

  const positionKey = getPositionKey(positionInfo)
  const hasMenuOpen = openMenuPositionKey === positionKey

  const onPress = useEvent(async () => {
    if (account.chainId !== chainId) {
      await switchChain(chainId)
    }

    accountDrawer.close()
    navigate(positionUrl)
  })

  const handleMenuOpenChange = useEvent((isOpen: boolean) => {
    setOpenMenuPositionKey(isOpen ? positionKey : null)
  })

  const analyticsEventProperties = useMemo(
    () => ({
      chain_id: chainId,
      pool_token_0_symbol: token0.symbol,
      pool_token_1_symbol: token1.symbol,
      pool_token_0_address: token0.isToken ? token0.wrapped.address : ZERO_ADDRESS,
      pool_token_1_address: token1.isToken ? token1.wrapped.address : ZERO_ADDRESS,
    }),
    [chainId, token0, token1],
  )

  return (
    <Trace logPress element={ElementName.MiniPortfolioPoolsRow} properties={analyticsEventProperties}>
      <TouchableArea onPress={onPress} mx="$spacing16" zIndex={hasMenuOpen ? zIndexes.mask : undefined}>
        <LiquidityPositionCard
          isMiniVersion
          liquidityPosition={positionInfo}
          showVisibilityOption
          isVisible={isVisible}
          onMenuOpenChange={handleMenuOpenChange}
        />
      </TouchableArea>
    </Trace>
  )
}
