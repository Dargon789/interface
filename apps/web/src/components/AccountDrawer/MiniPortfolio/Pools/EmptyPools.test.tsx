import { EmptyPools } from 'components/AccountDrawer/MiniPortfolio/Pools/EmptyPools'
import { render } from 'test-utils/render'

describe('EmptyPoolsModule', () => {
  it('matches base snapshot', () => {
    const { asFragment } = render(<EmptyPools />)
    expect(asFragment()).toMatchSnapshot()
  })
})
