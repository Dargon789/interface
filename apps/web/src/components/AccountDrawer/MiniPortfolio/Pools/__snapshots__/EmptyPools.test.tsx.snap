// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmptyPoolsModule > matches base snapshot 1`] = `
<DocumentFragment>
  <span
    class=""
    style="display: contents;"
  >
    <div
      class="_pt-_sm_t-space-spa94665593 _display-flex _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-column _alignItems-center _justifyContent-center _height-10037 _width-10037 _pl-t-space-spa1360334018 _pr-t-space-spa1360334018"
    >
      <svg
        fill="none"
        height="115"
        stroke-width="8"
        style="width: 115px; height: 115px; color: currentColor;"
        viewBox="0 0 115 115"
        width="115"
      >
        <path
          clip-rule="evenodd"
          d="M58.48 17C70.6302 17 80.48 26.8497 80.48 39C80.48 51.1503 70.6302 61 58.48 61C46.3297 61 36.48 51.1503 36.48 39C36.48 26.8497 46.3297 17 58.48 17ZM50.23 39L58.48 47.25L66.73 39L58.48 30.75L50.23 39Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
        <path
          d="M20 80.1986C30.4105 80.1986 38.2173 70.0581 38.2173 70.0581C38.2173 70.0581 46.0241 80.1986 56.4346 80.1986C66.8409 80.1986 77.2514 70.0581 77.2514 70.0581C77.2514 70.0581 87.6619 80.1986 95.4687 80.1986M20 99.2504C30.4105 99.2504 38.2173 89.1099 38.2173 89.1099C38.2173 89.1099 46.0241 99.2504 56.4346 99.2504C66.8409 99.2504 77.2514 89.1099 77.2514 89.1099C77.2514 89.1099 87.6619 99.2504 95.4687 99.2504"
          opacity="0.5"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="5"
        />
      </svg>
      <span
        class="font_subHeading _display-inline _boxSizing-border-box _whiteSpace-pre-wrap _mt-t-space-spa1360334080 _mr-0px _mb-0px _ml-0px _color-neutral1 _fontFamily-f-family _wordWrap-break-word _fontSize-f-size-larg101 _lineHeight-f-lineHeigh1262266530 _fontWeight-f-weight-bo3548 _textAlign-center"
        data-disable-theme="true"
      >
        No pools yet
      </span>
      <span
        class="font_body _textWrap-_platformweb_pretty _display-inline _boxSizing-border-box _whiteSpace-pre-wrap _mt-t-space-spa94665593 _mr-0px _mb-0px _ml-0px _color-neutral2 _fontFamily-f-family _wordWrap-break-word _fontSize-f-size-medi3736 _lineHeight-f-lineHeigh507465454 _fontWeight-f-weight-bo3548 _textAlign-center"
        data-disable-theme="true"
      >
        Open a new position or create a pool to get started.
      </span>
      <div
        class="_display-flex _alignItems-stretch _flexBasis-auto _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-0 _flexDirection-row _mt-t-space-spa1360334051"
      >
        <button
          class="is_Button _containerType-_platformweb_normal _outlineWidth-0focus-visible-1px _outlineOffset-0focus-visible-2px _outlineStyle-0focus-visible-solid _transform-0focus-visible-scaleX0--98sc575196563 _backgroundColor-0hover-accent1Hove112785 _backgroundColor-0focus-visible-accent1Hove112785 _outlineColor-0focus-visible-accent1Hove112785 _transform-0active-scale0--981281 _backgroundColor-0active-accent1Hove112785 _display-flex _flexBasis-0px _boxSizing-border-box _position-relative _minHeight-0px _minWidth-0px _flexShrink-1 _flexDirection-row _borderTopWidth-1px _borderRightWidth-1px _borderBottomWidth-1px _borderLeftWidth-1px _alignItems-center _justifyContent-center _backgroundColor-accent1 _borderTopColor-borderColor _borderRightColor-borderColor _borderBottomColor-borderColor _borderLeftColor-borderColor _cursor-pointer _height-auto _alignSelf-stretch _flexGrow-1 _pr-t-space-spa1360334080 _pl-t-space-spa1360334080 _pt-t-space-spa94665593 _pb-t-space-spa94665593 _borderTopLeftRadius-t-radius-ro291586453 _borderTopRightRadius-t-radius-ro291586453 _borderBottomRightRadius-t-radius-ro291586453 _borderBottomLeftRadius-t-radius-ro291586453 _gap-t-space-spa94665593 _borderBottomStyle-solid _borderTopStyle-solid _borderLeftStyle-solid _borderRightStyle-solid t_group_item"
          dd-action-name="+ New position"
        >
          <span
            class="font_button _display-inline _boxSizing-border-box _wordWrap-break-word _whiteSpace-nowrap _mt-0px _mr-0px _mb-0px _ml-0px _color-white _fontFamily-f-family _maxWidth-10037 _overflowX-hidden _overflowY-hidden _textOverflow-ellipsis _textAlign-center _fontSize-f-size-smal108 _fontWeight-f-weight-me3083741 _lineHeight-f-lineHeigh1269072494"
            line-height-disabled="false"
          >
            + New position
          </span>
        </button>
      </div>
    </div>
    <div
      style="display: contents;"
    />
  </span>
</DocumentFragment>
`;
